sage/group_prover.sage

```diff
@@ -0,0 +1,303 @@
+
+  def __div__(self,other):
+    """Divide two fractions."""
+    if parent(other) == ZZ:
+      return fastfrac(self.R,self.top,self.bot * other)
+    if other.__class__ == fastfrac:
+      return fastfrac(self.R,self.top * other.bot,self.bot * other.top)
+    return NotImplemented
+
+  def __pow__(self,other):
+    """Compute a power of a fraction."""
+    if parent(other) == ZZ:
+      if other < 0:
+        # Negative powers require flipping top and bottom
+        return fastfrac(self.R,self.bot ^ (-other),self.top ^ (-other))
+      else:
+        return fastfrac(self.R,self.top ^ other,self.bot ^ other)
+    return NotImplemented
+
+  def __str__(self):
+    return "fastfrac((" + str(self.top) + ") / (" + str(self.bot) + "))"
+  def __repr__(self):
+    return "%s" % self
+
+  def numerator(self):
+    return self.top
+
+def mynumerator(R,x):
+  """Compute the numerator of a fraction."""
+  if parent(x) == R:
+    return x
+  return numerator(x)
+
+class constraints:
+  def __init__(self, **kwargs):
+    if 'zero' in kwargs:
+      self.zero = dict(kwargs['zero'])
+    else:
+      self.zero = dict()
+    if 'nonzero' in kwargs:
+      self.nonzero = dict(kwargs['nonzero'])
+    else:
+      self.nonzero = dict()
+
+  def negate(self):
+    return constraints(zero=self.nonzero, nonzero=self.zero)
+
+  def __add__(self, other):
+    zero = self.zero.copy()
+    zero.update(other.zero)
+    nonzero = self.nonzero.copy()
+    nonzero.update(other.nonzero)
+    return constraints(zero=zero, nonzero=nonzero)
+
+  def __str__(self):
+    return "constraints(zero=%s,nonzero=%s)" % (self.zero, self.nonzero)
+
+  def __repr__(self):
+    return "%s" % self
+
+
+def conflicts(R, con):
+  """Check whether any of the passed non-zero assumptions is implied by the zero assumptions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), con.zero))
+  if len(con.nonzero) > 0 and 1 in zero:
+    return True
+  for nonzero in con.nonzero:
+    if nonzero.iszero(zero):
+      return True
+  if reduce(lambda a,b: a * b, con.nonzero, fastfrac(R, 1)).iszero(zero):
+    return True
+  return False
+
+
+def get_nonzero_set(R, assume):
+  """Calculate a simple set of nonzero expressions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), assume.zero))
+  nonzero = set()
+  for nz in assume.nonzero:
+    if mynumerator(R, nz) != 0:
+      for (f,n) in mynumerator(R, nz).factor():
+        nonzero.add(f)
+    rnz = zero.reduce(mynumerator(R, nz))
+    if rnz != 0:
+      for (f,n) in rnz.factor():
+        nonzero.add(f)
+  return nonzero
+
+
+def prove_nonzero(R, exprs, assume):
+  """Check whether an expression is provably nonzero, given assumptions"""
+  zero = R.ideal(map((lambda x: mynumerator(R, x)), assume.zero))
+  nonzero = get_nonzero_set(R, assume)
+  expl = set()
+  ok = True
+  for expr in exprs:
+    if mynumerator(R, expr) in zero:
+      return (False, [exprs[expr]])
+  allexprs = reduce(lambda a,b: a*b, exprs, fastfrac(R, 1))
```

Comment:
If you did `lambda a, b: numerator(a) * numerator(b)`, you could start with 1, and the function could take a list of numbers (rather than just fastfracs) as `expr`